### ISSUE TITLE
<pre> code blocks that contain HTML code don't display the HTML they contain

### ISSUE THREAD

**Issue by freakwriter** [wkhtmltopdf/wkhtmltopdf#4421]
**wkhtmltopdf version(s) affected**: 0.12.4

**OS information**
CentOS release 6.10

**Description**
We have a number of <pre> blocks that contain HTML code. <pre> blocks display fine when other code is displayed, but the blocks basically display as empty blocks if they contain only HTML, even when the HTML <> tags are encoded as &lt; and &gt;

**How to reproduce**
Here's two samples. The first is the the structure we're using on the HTML <pre> blocks (behavior is the same with and without the <code> sections) that don't display. Note the use of &lt; and &gt; within the pre blocks. It seems like wkhtmltopdf is parsing those as normal <> instead:
`<p>Here&#39;s some HTML with code tags:</p><pre><code data-language="html">&lt;body&gt;
&lt;p&gt;Here is some HTML.&lt;p&gt;
&lt;/body&gt;</code></pre><pre><code data-language="html">&lt;script&gt;
$(function(){

     var ToC =
     //Set the title for TOC. Can change h4 tags and text between them to change the text.
     //For no title, use: &quot;&lt;nav role=&#39;navigation&#39; class=&#39;table-of-contents toc-top&#39;&gt;&quot; + &quot;&lt;ul&gt;&quot;;
      &quot;&lt;nav role=&#39;navigation&#39; class=&#39;table-of-contents toc-top&#39;&gt;&lt;h4&gt;In this article:&lt;/h4&gt;&quot; + &quot;&lt;ul&gt;&quot;;
     var el, title, link;
      //Grab all the H2s and turn them into TOC titles. Adjust h2 in first line to use a different header level
     $(&quot;.hg-article-body h2&quot;).each(function() {
              el = $(this);
              title = el.text();
              anchorTitle = el.text().replace(/([~!@#$%^&amp;*()_+=`{}\[\]\|\\:;&#39;&lt;&gt;,.\/\? ])+/g, &#39;-&#39;).toLowerCase();
              link = &quot;#&quot; + anchorTitle;
              el.html(&#39;&lt;a id=&quot;&#39;+anchorTitle+&#39;&quot; class=&quot;toc-anchor&quot;&gt;&lt;/a&gt;&#39; + el.html());
              newLine =
    &quot;&lt;li&gt;&quot; +
      &quot;&lt;a class=&#39;article-anchor&#39; href=&#39;&quot; + link + &quot;&#39;&gt;&quot; +
        title +
      &quot;&lt;/a&gt;&quot; +
    &quot;&lt;/li&gt;&quot;;

  ToC += newLine;
     });
     ToC +=
   &quot;&lt;/ul&gt;&quot; +
  &quot;&lt;/nav&gt;&quot;;
    $(&quot;#snippet-prepend&quot;).before(ToC);   
});

&lt;/script&gt;</code></pre>`

But <pre> blocks work when HTML isn't involved, as in this sample:
`<p>If I do straight Javascript with no HTML tags, it renders (with and without code tags)</p>
<pre><code data-language="javascript">
$(function(){
     var ToC =
     //Set the title for TOC. Can change h4 tags and text between them to change the text.
     //For no title, use: &quot;&lt;nav role=&#39;navigation&#39; class=&#39;table-of-contents toc-top&#39;&gt;&quot; + &quot;&lt;ul&gt;&quot;;
      &quot;&lt;nav role=&#39;navigation&#39; class=&#39;table-of-contents toc-top&#39;&gt;&lt;h4&gt;In this article:&lt;/h4&gt;&quot; + &quot;&lt;ul&gt;&quot;;
     var el, title, link;
      //Grab all the H2s and turn them into TOC titles. Adjust h2 in first line to use a different header level
     $(&quot;.hg-article-body h2&quot;).each(function() {
              el = $(this);
              title = el.text();
              anchorTitle = el.text().replace(/([~!@#$%^&amp;*()_+=`{}\[\]\|\\:;&#39;&lt;&gt;,.\/\? ])+/g, &#39;-&#39;).toLowerCase();
              link = &quot;#&quot; + anchorTitle;
              el.html(&#39;&lt;a id=&quot;&#39;+anchorTitle+&#39;&quot; class=&quot;toc-anchor&quot;&gt;&lt;/a&gt;&#39; + el.html());
              newLine =
    &quot;&lt;li&gt;&quot; +
      &quot;&lt;a class=&#39;article-anchor&#39; href=&#39;&quot; + link + &quot;&#39;&gt;&quot; +
        title +
      &quot;&lt;/a&gt;&quot; +
    &quot;&lt;/li&gt;&quot;;

  ToC += newLine;
     });
     ToC +=
   &quot;&lt;/ul&gt;&quot; +
  &quot;&lt;/nav&gt;&quot;;
    $(&quot;#snippet-prepend&quot;).before(ToC);   
});
</code></pre>`

**Expected behavior**
I would expect that the encoding of the  < > within <pre> blocks would prevent the included HTML from being parsed as straight HTML--the text would display in the pre blocks as it does in a browser. (While allowing proper HTML parsing of things still in legit <tags>, like <code>.)

Didn't see this anywhere else so perhaps it's something unique to our setup...?

**Comment by ashkulz** [comment]
Weird, is this an issue with the 0.13 alpha builds?

**Comment by freakwriter** [comment]
@ashkulz no--we've seen this behavior in both 0.12.3 and 0.12.4. Any ideas what we might be doing wrong, if this isn't a known issue?

**Comment by ashkulz** [comment]
I think this is possibly an issue with the layout engine in 0.12 -- it's quite old, and looks to be a bug in that I'm afraid.

**Comment by freakwriter** [comment]
Ah, okay. At least it's not just us.

Do you have a timeline as to when 0.13 will be done with beta and become
the current stable version?

**Comment by ashkulz** [comment]
There's no timeline -- it's an open source project :slightly_smiling_face: -- I'm hoping to do it in the next 4-5 months.

**Comment by freakwriter** [comment]
@ashkulz I figured that was the situation, but just thought I'd ask!

I'm hesitant to go down the path of testing the alpha builds, but could you include this test case in the 0.13 builds? Would be great if it worked!

**Comment by PhilterPaper** [comment]
Just in case anyone reading this is confused (or, it's possible that _I_ am!), my understanding of `<pre>` is that it is block level, normally the font is switched to a monospaced font, and line reflow is shut off (whitespace runs are not collapsed to a single space, and lines are not jammed together). Semantically, it really doesn't mean anything. `<code>` is inline and normally just switches the font to a monospaced font. Semantically, it can be used to indicate that this text is special (that it is computer code, not English prose). In **both** cases, HTML-like markup within them (i.e., within `<>`) _should_ be normally processed (e.g., `<b>` should cause **bold** text). I'm not sure that nesting `<code>` within `<pre>` does anything useful (unless `<pre>` has been broken to not switch to a monospaced font). Just for grins and giggles, see if removing `<code>` makes any difference.

Changing `<` and `>` to & lt; and & gt; _should_ disable these tags (it does in a browser). If wkHTMLtoPDF is replacing the entities by `<` and `>` for printing purposes, and _then_ interprets them as HTML tags, that sounds like an error. It would be interesting to know if the fault is in WebKit, or if somewhere in wkHTMLtoPDF it is swapping entities for characters too early.

**Comment by ferrall** [comment]
[Windows]
I have tried several ways to get monospace fonts into a wkhtmltopdf PDF in a large HTML document with a lot of TeX math rendered using Mathjax.  <code>, <pre> and font-family in CSS style all fail to change to monospace.    
 
The problem goes away when I eliminate MathJax script to render the math.  And it goes away if I keep MathJax but I drastically reduce the amount of TeX code to process. 

if I switch the MathJax configuration from CHTML to MML_HTMLorMML then <code> starts to work but not <pre>.  So there seems to be some interaction between the MathJax script and wkhtmltopdf that I can modified but not in a predictable way.  Frustrating ...
